modules/tinymce/src/core/main/ts/mode/Readonly.ts

```diff
@@ -38,23 +39,54 @@ const registerReadOnlyInputBlockers = (editor: Editor): void => {
     }
   });
 
-  editor.on('input', (e) => {
-    if (!e.isComposing && isReadOnly(editor)) {
-      const undoLevel = editor.undoManager.add();
-      if (Type.isNonNullable(undoLevel)) {
-        editor.undoManager.undo();
+  // Prevent undo level additions during composition
+  editor.on('input compositionend', (e) => {
+    if (isReadOnly(editor)) {
+      // Skip undo level addition during composition
+      if (!e.isComposing) {
+        const undoLevel = editor.undoManager.add();
+        if (Type.isNonNullable(undoLevel)) {
+          editor.undoManager.undo();
+        }
       }
     }
   });
 
-  editor.on('compositionend', () => {
+  // Set up mutation observer to detect and revert unintended changes
+  const observer = new MutationObserver((mutations) => {
     if (isReadOnly(editor)) {
-      const undoLevel = editor.undoManager.add();
-      if (Type.isNonNullable(undoLevel)) {
-        editor.undoManager.undo();
+      let needsRevert = false;
+
+      mutations.forEach((mutation) => {
+        if (mutation.type === 'characterData' || mutation.type === 'childList') {
+          // Mark that we need to revert
+          needsRevert = true;
+        }
+      });
+
+      // Only create one undo level for all mutations in this batch
+      if (needsRevert) {
+        const undoLevel = editor.undoManager.add();
```

Comment:
Since this is called two times move it to a function that captures intent like `rollbackUndo` or something.
Also when you `undo` the stack still has the extra `redo` level so ideally we should try to get rid of that since this would cause weird effects if you say do a mutation in readonly mode. Then toggle readonly off then you have a `redo` state that you can move to unless I'm missing something here.